core/shared/src/main/scala/org/http4s/Headers.scala

```diff
@@ -70,15 +70,73 @@ final class Headers(val headers: List[Header.Raw]) extends AnyVal {
     * @return a new [[Headers]] containing the sum of the initial and input headers
     */
   def put(in: Header.ToRaw*): Headers =
-    this ++ Headers(in.values)
+    in match {
+      case thatHeader :: Nil =>
+        putOne(thatHeader)
+      case Nil =>
+        this
+      case _ =>
+        this ++ Headers(in.values)
+    }
+
+  private def putOne(header: Header.ToRaw): Headers =
+    if (this.headers.isEmpty) {
+      Headers(header)
+    } else {
+      header.values match {
+        case values @ head :: _ =>
+          val thatHeaderName = head.name
+          if (!headers.exists(_.name == thatHeaderName)) {
+            Headers(this.headers ++ values)
+          } else {
+            val newHeaders = mutable.ListBuffer.empty[Header.Raw]
+            headers.foreach { h =>
+              if (h.name != thatHeaderName) {
+                newHeaders += h
+              }
+            }
+            newHeaders ++= values
+            Headers(newHeaders.toList)
+          }
+        // Shouldn't really happen, but if it does, there's nothing to be done.
+        case Nil =>
+          this
+      }
+    }
 
   def ++(those: Headers): Headers =
     if (those.headers.isEmpty) this
     else if (this.headers.isEmpty) those
     else {
-      val thoseNames = mutable.Set.empty[CIString]
-      those.headers.foreach(h => thoseNames.add(h.name))
-      Headers(headers.filterNot(h => thoseNames.contains(h.name)) ++ those.headers)
+      those.headers match {
+        case thatHeader :: Nil =>
+          val thatHeaderName = thatHeader.name
+          if (!headers.exists(_.name == thatHeader.name)) {
+            Headers(this.headers :+ thatHeader)
+          } else {
+            val newHeaders = mutable.ListBuffer.empty[Header.Raw]
+            headers.foreach { h =>
+              if (h.name != thatHeaderName) {
+                newHeaders += h
+              }
+            }
+            newHeaders += thatHeader
+            Headers(newHeaders.toList)
+          }
```

Comment:
Is it just me or is the `!headers.exists(...)` doing an extra traversal? It seems like no matter whether the result is `true` or `false` we end up creating a new list and appending to the end.

Reply:
It does, that's 3. from my original comment:

3. We could keep track of header names in a mutable set internally, which would remove the need for an extra traversal in `!headers.exists(_.name == thatHeaderName)`, but would that be worth doing, in terms of added code, memory usage, and mutable state? I haven't done any benchmarks on this, yet.

Reply:
My gut feeling is that it would be worth it if we insist on staying with `List`, but not worth it if we want change the underlying data structure to something more appropriate, and I'm strongly in favor of the latter.

Reply:
Hum, sorry, I'm confused. Can't we write this as equivalently like this? Apologies if I am being stupid 😅 
```suggestion
        case thatHeader :: Nil =>
          val thatHeaderName = thatHeader.name
          val newHeaders = mutable.ListBuffer.empty[Header.Raw]
          headers.foreach { h =>
            if (h.name != thatHeaderName) {
              newHeaders += h
            }
          }
          newHeaders += thatHeader
          Headers(newHeaders.toList)
```

Reply:
Although it's late over here, you're definitely not stupid, and it looks to me like that would work. The tests agree. It might have been more reasonable when doing preprends. :thinking:

Reply:
Yes, when `prepend`ing you wouldn't have to copy the list. But in this case we're copying no matter what.